src/protocol/loli_nrf24l01.c

```diff
@@ -0,0 +1,420 @@
+    CHANNEL3,
+    CHANNEL4,
+    CHANNEL5,
+    CHANNEL6,
+    CHANNEL7,
+    CHANNEL8,
+};
+
+// Bit vector from bit position
+#define BV(bit) (1 << bit)
+
+static void init_RF()
+{
+    NRF24L01_Initialize();
+    NRF24L01_FlushTx();
+    NRF24L01_FlushRx();
+    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);     // Clear data ready, data sent, and retransmit
+    NRF24L01_WriteReg(NRF24L01_01_EN_AA, 0x00);      // No Auto Acknowldgement on all data pipes
+    NRF24L01_WriteReg(NRF24L01_02_EN_RXADDR, 0x01);  // enable rx data pipe 0
+    NRF24L01_WriteReg(NRF24L01_04_SETUP_RETR, 0x00);    // No retransmit
+    NRF24L01_WriteReg(NRF24L01_05_RF_CH, 66);   // start frequency
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, LOLI_PACKET_SIZE);  // RX FIFO size
+    NRF24L01_SetBitrate(NRF24L01_BR_250K);
+    tx_power = Model.tx_power;
+    NRF24L01_SetPower(tx_power);
+    NRF24L01_SetTxRxMode(TX_EN);
+    NRF24L01_WriteReg(NRF24L01_00_CONFIG, 0x0a);  // 8bit CRC, TX
+}
+
+#define CHAN_RANGE (CHAN_MAX_VALUE - CHAN_MIN_VALUE)
+static u16 scale_channel(u8 ch, u16 destMin, u16 destMax)
+{
+    s32 chanval = Channels[ch];
+    s32 range = (s32) destMax - (s32) destMin;
+
+    if (chanval < CHAN_MIN_VALUE)
+        chanval = CHAN_MIN_VALUE;
+    else if (chanval > CHAN_MAX_VALUE)
+        chanval = CHAN_MAX_VALUE;
+    return (range * (chanval - CHAN_MIN_VALUE)) / CHAN_RANGE + destMin;
+}
+
+static u8 rxConfigChanged()
+{
+    u16 temp = 0;
+    for (u8 i=0; i < 8; i++)
+        temp += Model.proto_opts[PROTOOPTS_CH1 + i];
+    if (temp != rx_config) {
+        rx_config = temp;
+        return 1;
+    }
+    return 0;
+}
+
+static void send_rx_config()
+{
+    packet[0] = 0xa2;
+    // default = set all outputs to servo mode
+    packet[1] = 0;
+    packet[2] = 0;
+    // Rx Output #1 PPM or PWM
+    if (Model.proto_opts[PROTOOPTS_CH1] == OPT_PPM)
+        packet[1] |= FLAG_PPM;
+    else if (Model.proto_opts[PROTOOPTS_CH1] == OPT_PWM)
+        packet[1] |= FLAG_PWM1;
+    // Rx Output #2 PWM
+    if (Model.proto_opts[PROTOOPTS_CH2] == OPT_PWM)
+        packet[1] |= FLAG_PWM2;
+    // Rx Output #5 SBUS
+    if (Model.proto_opts[PROTOOPTS_CH5] == OPT_SBUS)
+        packet[1] |= FLAG_SBUS;
+    // Rx Output #7 PWM
+    if (Model.proto_opts[PROTOOPTS_CH7] == OPT_PWM)
+        packet[1] |= FLAG_PWM7;
+    // Rx Output X "Switch" mode
+    for (u8 i=0; i < 8; i++) {
+        if (Model.proto_opts[PROTOOPTS_CH1 + i] == OPT_SW)
+            packet[2] |= 1 << (7-i);
+    }
+
+    if (++hopping_frequency_no > LOLI_NUM_CHANNELS-1)
+        hopping_frequency_no = 0;
+    NRF24L01_WriteReg(NRF24L01_05_RF_CH, hopping_frequency[hopping_frequency_no]);
+    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);
+    NRF24L01_FlushTx();
+    NRF24L01_WritePayload(packet, LOLI_PACKET_SIZE);
+}
+
+static void send_packet(u8 bind)
+{
+    u16 val;
+    if (bind) {
+        packet[0] = 0xa0;
+        memcpy(&packet[1], hopping_frequency, 5);
+        memcpy(&packet[6], rx_tx_addr, 5);
```

Comment:
why not move set channel to 33 here? so two code pathes of send packet are symmetry.